src/main/kotlin/com/openlattice/chronicle/services/jobs/JobService.kt

```diff
@@ -0,0 +1,130 @@
+import com.geekbeast.postgres.PostgresArrays
+import com.geekbeast.postgres.PostgresDatatype
+import com.geekbeast.postgres.streams.BasePostgresIterable
+import com.geekbeast.postgres.streams.PreparedStatementHolderSupplier
+import com.geekbeast.rhizome.jobs.JobStatus
+import com.openlattice.chronicle.storage.StorageResolver
+import org.springframework.stereotype.Service
+import com.openlattice.chronicle.ids.HazelcastIdGenerationService
+import com.openlattice.chronicle.jobs.ChronicleJob
+import com.openlattice.chronicle.postgres.ResultSetAdapters
+import com.openlattice.chronicle.storage.ChroniclePostgresTables.Companion.JOBS
+import com.openlattice.chronicle.storage.PostgresColumns
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.CONTACT
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.DELETED_ROWS
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.JOB_DEFINITION
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.JOB_ID
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.MESSAGE
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.PRINCIPAL_ID
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.PRINCIPAL_TYPE
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.SECURABLE_PRINCIPAL_ID
+import com.openlattice.chronicle.storage.PostgresColumns.Companion.STATUS
+import org.slf4j.LoggerFactory
+import java.sql.Connection
+import java.util.*
+import java.util.concurrent.ConcurrentSkipListMap
+
+/**
+ * @author Solomon Tang <solomon@openlattice.com>
+ */
+@Service
+class JobService(
+    private val idGenerationService: HazelcastIdGenerationService,
+    private val storageResolver: StorageResolver,
+) : JobManager {
+    companion object {
+        private val logger = LoggerFactory.getLogger(JobService::class.java)
+        private val mapper = ObjectMappers.newJsonMapper()
+        private val running = ConcurrentSkipListMap<UUID, ChronicleJob>()
+
+        private val JOB_COLUMNS_LIST = listOf(
+            JOB_ID,
+            SECURABLE_PRINCIPAL_ID,
+            PRINCIPAL_TYPE,
+            PRINCIPAL_ID,
+            STATUS,
+            CONTACT,
+            JOB_DEFINITION,
+            MESSAGE,
+            DELETED_ROWS,
+        )
+        private val JOB_COLUMNS = JOB_COLUMNS_LIST.joinToString(",") { it.name }
+        private val JOB_COLUMNS_BIND = JOB_COLUMNS_LIST.joinToString(",") {
+            if (it.datatype == PostgresDatatype.JSONB) "?::jsonb" else "?"
+        }
+        private val INSERT_JOB_SQL = """
+            INSERT INTO ${JOBS.name} ($JOB_COLUMNS) VALUES ($JOB_COLUMNS_BIND)
+        """.trimIndent()
+
+        private val GET_JOBS_SQL = """
+            SELECT * FROM ${JOBS.name} WHERE ${JOB_ID.name} = ANY(?)
+        """.trimIndent()
+
+        private val GET_NEXT_JOB_SQL = """
+            UPDATE ${JOBS.name}
+            SET ${STATUS.name} = '${JobStatus.FINISHED.name}'
```

Comment:
Why set to finished?

Reply:
When the audited transaction for the job completes, this includes any calls to redshift to remove usage data. If this job does not finish, then it will automatically rollback to it's PENDING state to be picked up by another worker if necessary

Reply:
Wouldn't it make more sense to do it when `updateFinishedDeleteJob` is called though, which already does other final updates? I think it's dangerous to have a `GET_SOMETHING_SQL` constant that does `UPDATE`, feels like it could easily be misused.